File: 2 SEGUNDO/1 Cuatrimestre/Programación y Diseño Orientado a Objetos - PDOO/Practicas/rubyNapakalaki/card_dealer.rb

require_relative 'prize.rb'
require_relative 'bad_consequence.rb'
require_relative 'monster.rb'
require_relative 'treasure_kind.rb'
require_relative 'cultist.rb'

require 'singleton'
class CardDealer

  include Singleton
  attr_accessor :unused_treasures ,:used_treasures, :unused_monster ,:used_monster, :used_cultist, :unused_cultist

  def initialize

  end

  def self.get_instance
    self.instance
  end

  def init_treasure_card_deck
      @unused_treasures = Array.new
      @used_treasures = Array.new
      @unused_treasures << Treasure.new("¡Si mi amo!", 4, [TreasureKind::HELMET])
      @unused_treasures << Treasure.new("Botas de investigación", 3, [TreasureKind::SHOES])
      @unused_treasures << Treasure.new("Capucha de Cthulhu", 3, [TreasureKind::HELMET])
      @unused_treasures << Treasure.new("A prueba de babas", 2, [TreasureKind::ARMOR])
      @unused_treasures << Treasure.new("Botas de lluvia ácida", 1, [TreasureKind::BOTHHANDS])
      @unused_treasures << Treasure.new("Casco minero", 2, [TreasureKind::HELMET])
      @unused_treasures << Treasure.new("Ametralladora ACME", 4, [TreasureKind::BOTHHANDS])
      @unused_treasures << Treasure.new("Camiseta de la ETSIIT", 1, [TreasureKind::ARMOR])
      @unused_treasures << Treasure.new("Clavo de rail ferroviario", 3, [TreasureKind::ONEHAND])
      @unused_treasures << Treasure.new("Cuchillo de sushi arcano", 2, [TreasureKind::ONEHAND])
      @unused_treasures << Treasure.new("Fez alópodo", 3, [TreasureKind::HELMET])
      @unused_treasures << Treasure.new("Hacha prehistórica", 2, [TreasureKind::ONEHAND])
      @unused_treasures << Treasure.new("El aparato del Pr. Tesla", 4, [TreasureKind::ARMOR])
      @unused_treasures << Treasure.new("Gaita", 4, [TreasureKind::BOTHHANDS])
      @unused_treasures << Treasure.new("Insecticida", 2, [TreasureKind::ONEHAND])
      @unused_treasures << Treasure.new("Escopeta de 3 cañones", 4, [TreasureKind::BOTHHANDS])
      @unused_treasures << Treasure.new("Garabato Mistico", 2, [TreasureKind::ONEHAND])
      @unused_treasures << Treasure.new("La rebeca metálica", 2, [TreasureKind::ARMOR])
      @unused_treasures << Treasure.new("Mazo de los antiguos", 3, [TreasureKind::ONEHAND])
      @unused_treasures << Treasure.new("Lanzallamas", 4, [TreasureKind::BOTHHANDS])
      @unused_treasures << Treasure.new("Necro-comicón", 1, [TreasureKind::ONEHAND])
      @unused_treasures << Treasure.new("Necronomicón", 5, [TreasureKind::BOTHHANDS])
      @unused_treasures << Treasure.new("Linterna a 2 manos", 3, [TreasureKind::BOTHHANDS])
      @unused_treasures << Treasure.new("Necro-gnomicón", 2, [TreasureKind::ONEHAND])
      @unused_treasures << Treasure.new("Necrotelecom", 2, [TreasureKind::HELMET])
      @unused_treasures << Treasure.new("Mazo de los antiguos", 3, [TreasureKind::ONEHAND])
      @unused_treasures << Treasure.new("Necro-playboycon", 3, [TreasureKind::ONEHAND])
      @unused_treasures << Treasure.new("Porra preternatural", 2, [TreasureKind::ONEHAND])
      @unused_treasures << Treasure.new("Shogulador", 1, [TreasureKind::BOTHHANDS])
      @unused_treasures << Treasure.new("Varita de atizamiento", 3, [TreasureKind::ONEHAND])
      @unused_treasures << Treasure.new("Tentácula de pega", 2, [TreasureKind::HELMET])
      @unused_treasures << Treasure.new("Zapato deja-amigos", 1, [TreasureKind::SHOES])


  end
  private :init_treasure_card_deck

  def init_monster_card_deck
    @unused_monster = Array.new
    @used_monster = Array.new
    
     

    prize1 = Prize.new(2,1)

    bad_consequence1 = SpecificBadConsequence.new('Pierdes tu armadura visible y otra oculta.', 0 ,[TreasureKind::ARMOR],[TreasureKind::ARMOR])

    @unused_monster << Monster.new('Byakhees de bonanza', 8, bad_consequence1, prize1,0)

    

    prize2 = Prize.new(1,1)

    bad_consequence2 = SpecificBadConsequence.new('Embobados con el lindo primigenio te descartas de tu casco visible.', 0 ,[TreasureKind::HELMET],[])

    @unused_monster << Monster.new('Tenochtitlan', 2, bad_consequence2, prize2,0)

    

    prize3 = Prize.new(1,1)

    bad_consequence3 = SpecificBadConsequence.new('El primordial bostezo contagioso. Pierdes el calzado visible.', 0 ,[TreasureKind::SHOES],[])

    @unused_monster << Monster.new('EL sopor de Dunwich', 2, bad_consequence3, prize3,0)


    

    prize4 = Prize.new(4,1)

    bad_consequence4 = SpecificBadConsequence.new('Te atrapan para llevarte de fiesta y te dejan caer en mitad del vuelo. Descarta 1 mano visible y 1 mano oculta.', 0,
      [TreasureKind::ONEHAND],[TreasureKind::ONEHAND])

    @unused_monster << Monster.new('Angeles de la noche ibicenca', 14, bad_consequence4, prize4,0)

    

    prize5 = Prize.new(3,1)

    bad_consequence5 = NumericBadConsequence.new('Pierdes todos tus tesoros visibles.', 0 ,6,0)

    @unused_monster << Monster.new('El gorron en el umbral', 10, bad_consequence5, prize5,0)

    

    prize6 = Prize.new(2,1)

    bad_consequence6 = SpecificBadConsequence.new('Pierdes tu armadura visible.', 0, [TreasureKind::ARMOR],[])

    @unused_monster << Monster.new('H.P. Munchcraft', 6, bad_consequence6, prize6,0)

    

    prize7 = Prize.new(1,1)

    bad_consequence7 = SpecificBadConsequence.new('Sientes bichos bajo la ropa. Descartas la armadura visible.', 0, [TreasureKind::ARMOR],[])

    @unused_monster << Monster.new('Necrofago', 13, bad_consequence7, prize7,0)


    

    prize8 = Prize.new(4,2)

    bad_consequence8 = NumericBadConsequence.new('Pierdes 5 niveles y 3 tesoros visibles',5 , 3, 0)

    @unused_monster << Monster.new('El rey de rosa',13 , bad_consequence8 , prize8,0)

    

    prize9 = Prize.new(1,1)

    bad_consequence9 = NumericBadConsequence.new('Toses los pulmones y pierdes 2 niveles.',2 ,0, 0)

    @unused_monster << Monster.new('Flecher',2 , bad_consequence9 , prize9,0)

    

    prize10 = Prize.new(4,2)

    bad_consequence10 = DeathBadConsequence.new('Estos monstruos resultan bastantes superficiales y te aburren mortalmente. Estas muerto.')

    @unused_monster << Monster.new('Los hondos',8 , bad_consequence10 , prize10,0)

    

    prize11 = Prize.new(2,1)

    bad_consequence11 = NumericBadConsequence.new('Pierdes 2 niveles y 2 tesoros ocultos.',2 , 0, 2)

    @unused_monster << Monster.new('Semillas Cthulhu',4, bad_consequence11 , prize11,0)

    

    prize12 = Prize.new(2,1)

    bad_consequence12 = SpecificBadConsequence.new('Te intentas escaquear. Pierdes una mano visible.', 0, [TreasureKind::ONEHAND],[])

    @unused_monster << Monster.new('Dameargo',1, bad_consequence12 , prize12,0)

    

    prize13 = Prize.new(1,1)

    bad_consequence13 = NumericBadConsequence.new('Da mucho asquito. Pierdes 3 niveles.',3 , 0, 0)

    @unused_monster << Monster.new('Pollipolipo volante',3, bad_consequence13 , prize13,0)

    

    prize14 = Prize.new(3,1)

    bad_consequence14 = DeathBadConsequence.new('No le hace gracia que pronuncien mal su nombre. Estas muerto.')

    @unused_monster << Monster.new('Yskhtihyssg-Goth',12 , bad_consequence14 , prize14,0)

    

    prize15 = Prize.new(4,1)

    bad_consequence15 = DeathBadConsequence.new('La familia te atrapa. Estas muerto.')

    @unused_monster << Monster.new('Familia feliz',1 , bad_consequence15 , prize15,0)

    

    prize16 = Prize.new(2,1)

    bad_consequence16 = SpecificBadConsequence.new('La quinta directiva te obliga a perder 2 niveles y un tesoro de 2 manos visible.', 2, [TreasureKind::BOTHHANDS],[])

    @unused_monster << Monster.new('Roboggoth',8, bad_consequence16 , prize16,0)

    

    prize17 = Prize.new(1,1)

    bad_consequence17 = SpecificBadConsequence.new('Te asusta en la noche. Pierdes un casco visible.', 0, [TreasureKind::HELMET],[])

    @unused_monster << Monster.new('El espia sordo',5, bad_consequence17 , prize17,0)

    

    prize18 = Prize.new(1,1)

    bad_consequence18 = NumericBadConsequence.new('Menudo susto te llevas.Pierdes 2 niveles y 5 tesoros visibles.',2 , 5, 0)

    @unused_monster << Monster.new('Tongue',19, bad_consequence18 , prize18,0)

    

    prize19 = Prize.new(1,1)

    bad_consequence19 = SpecificBadConsequence.new('Te faltan manos para tanta cabeza. Pierdes todos tus tesoros visibles de las manos.', 0,
                            [TreasureKind::BOTHHANDS,TreasureKind::ONEHAND,TreasureKind::ONEHAND],[])

    @unused_monster << Monster.new('Bicefalo',21, bad_consequence19 , prize19,0)


    
       bad_consequence20 = SpecificBadConsequence.new('Pierdes una mano visible' ,0 ,
               [TreasureKind::ONEHAND],[])
       prize_20 = Prize.new(3,1)

      @unused_monster << Monster.new('El mal indecible impronunciable.' ,10 , bad_consequence20, prize_20, -2)

       
       bad_consequence21 =  NumericBadConsequence.new('Pierdes tus tesoros visibles. Jajaja',0 ,5,0)
       prize_21 = Prize.new(2,1)
       @unused_monster << Monster.new('Testigos Oculares.' ,6 , bad_consequence21, prize_21, +2)

       
       bad_consequence22 = DeathBadConsequence.new('Hoy no es tu dia de suerte. Mueres.')
       prize_22 = Prize.new(2,5)
       @unused_monster << Monster.new('El gran cthulhu.',20 , bad_consequence22, prize_22,+4)

       
       bad_consequence23 = NumericBadConsequence.new('Tu gobierno te recorta dos niveles.' ,2 ,0,0)
       prize_23 = Prize.new(2,1)
       @unused_monster << Monster.new('Serpiente politico.' ,8 , bad_consequence23, prize_23, -2)

       
       bad_consequence24 = SpecificBadConsequence.new('Pierdes tu casco y armadura visible. Pierdes tus manos ocultas.' ,0 ,
               [TreasureKind::HELMET,TreasureKind::ARMOR],[TreasureKind::BOTHHANDS,TreasureKind::ONEHAND,TreasureKind::ONEHAND])
       prize_24  = Prize.new(1,1)
       @unused_monster << Monster.new('Felpuggoth.' ,2 , bad_consequence24, prize_24, +5)

       
       bad_consequence25 = NumericBadConsequence.new('Pierdes dos niveles.' ,2 ,0,0)
       prize_25 = Prize.new(4,2)
       @unused_monster <<  Monster.new('Shoggoth.' ,16 , bad_consequence25, prize_25, -4)

       
       bad_consequence26 = NumericBadConsequence.new('Pintalabios negro. Pierdes dos niveles.' ,2 ,0,0)
       prize_26 = Prize.new(1,1)
       @unused_monster << Monster.new('Lolitaggoth.' ,2 , bad_consequence26, prize_26, +3)

    

  end

  private :init_monster_card_deck



   def init_cultist_card_deck
    @unused_cultist = Array.new
    @unused_cultist << Cultist.new('Sectario', 1)
    @unused_cultist << Cultist.new('Sectario', 2)
    @unused_cultist << Cultist.new('Sectario', 1)
    @unused_cultist << Cultist.new('Sectario', 2)
    @unused_cultist << Cultist.new('Sectario', 1)
    @unused_cultist << Cultist.new('Sectario', 1)

    end

   private :init_cultist_card_deck


   def shuffle_cultist
    @unused_cultist = @unused_cultist.shuffle
   end

  def shuffle_treasures
    @unused_treasures = @unused_treasures.shuffle
  end

  def shuffle_monster
    @unused_monster = @unused_monster.shuffle
  end

  def next_cultist

      if @unused_cultist.empty?

        @used_cultist.each do |cultist|

          @unused_cultist<<cultist

        end

        shuffle_cultist
        @used_cultist.clear

      end

      cultist = @unused_cultist.at(0)

      @used_cultist<<cultist

      @unused_cultist.delete(cultist)
      cultist

  end

  def next_treasure

      if @unused_treasures.empty?

        @used_treasures.each do |treasure|

          @unused_treasures<<treasure

        end

        shuffle_treasures
        @used_treasures.clear

      end

      treasure = @unused_treasures.at(0)

      @used_treasures<<treasure

      @unused_treasures.delete(treasure)
      treasure

  end

  def next_monster
    if @unused_monster.empty?

        @used_monster.each do |monster|

          @unused_monster<<monster

        end

        shuffle_monster
        @used_monster.clear

      end

      monster = @unused_monster.at(0)

      @used_monster<<monster

      @unused_monster.delete(monster)
      monster

  end

  def give_cultist_back(cultist)
    @used_cultist<< cultist
  end

  def give_treasure_back(treasure)
    @used_treasures << treasure
  end

  def give_monster_back(monster)
    @used_monster << monster
  end

  def init_cards
    init_cultist_card_deck
    init_treasure_card_deck
    init_monster_card_deck
    shuffle_cultist
    shuffle_treasures
    shuffle_monster
  end

end

File: 2 SEGUNDO/1 Cuatrimestre/Programación y Diseño Orientado a Objetos - PDOO/Practicas/javaNapakalaki/NapakalakiGame/CardDealer.java
package NapakalakiGame;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;

/**
 *
 * @author manjaro
 */

public class CardDealer {

      private static final CardDealer instance = new CardDealer();
      private ArrayList<Treasure> unusedTreasures ;
      private ArrayList<Treasure> usedTreasures ;
      private ArrayList<Monster> unusedMonster ;
      private ArrayList<Monster> usedMonster ;

      private ArrayList<Cultist> unusedCultist;
      private ArrayList<Cultist> usedCultist;


   public static CardDealer getInstance() {
      return instance;
  }


  private CardDealer(){

      unusedMonster = new ArrayList();
      usedMonster = new ArrayList();
      unusedTreasures = new ArrayList();
      usedTreasures = new ArrayList();
      unusedCultist = new ArrayList();
      usedCultist = new ArrayList();
  }
  private void initTreasureCardDeck(){

    unusedTreasures.add(new Treasure("¡Si mi amo!", 4, TreasureKind.HELMET));
    unusedTreasures.add(new Treasure("Botas de investigación", 3, TreasureKind.SHOES));
    unusedTreasures.add(new Treasure("Capucha de Cthulhu", 3 , TreasureKind.HELMET));
    unusedTreasures.add(new Treasure("A prueba de babas", 2, TreasureKind.ARMOR));
    unusedTreasures.add(new Treasure("Botas de lluvia ácida", 1, TreasureKind.BOTHHANDS));
    unusedTreasures.add(new Treasure("Casco minero", 2, TreasureKind.HELMET));
    unusedTreasures.add(new Treasure("Ametralladora ACME", 4, TreasureKind.BOTHHANDS));
    unusedTreasures.add(new Treasure("Camiseta de la ETSIIT", 1, TreasureKind.ARMOR));
    unusedTreasures.add(new Treasure("Clavo de rail ferroviario", 3, TreasureKind.ONEHAND));
    unusedTreasures.add(new Treasure("Cuchillo de sushi arcano", 2, TreasureKind.ONEHAND));
    unusedTreasures.add(new Treasure("Fez alópodo", 3, TreasureKind.HELMET));
    unusedTreasures.add(new Treasure("Hacha prehistórica", 2, TreasureKind.ONEHAND));
    unusedTreasures.add(new Treasure("El aparato del Pr. Tesla", 4, TreasureKind.ARMOR));
    unusedTreasures.add(new Treasure("Gaita", 4, TreasureKind.BOTHHANDS));
    unusedTreasures.add(new Treasure("Insecticida", 2, TreasureKind.ONEHAND));
    unusedTreasures.add(new Treasure("Escopeta de 3 cañones", 4, TreasureKind.BOTHHANDS));
    unusedTreasures.add(new Treasure("Garabato Mistico", 2, TreasureKind.ONEHAND));
    unusedTreasures.add(new Treasure("La rebeca metálica", 2, TreasureKind.ARMOR));
    unusedTreasures.add(new Treasure("Mazo de los antiguos", 3, TreasureKind.ONEHAND));
    unusedTreasures.add(new Treasure("Lanzallamas", 4, TreasureKind.BOTHHANDS));
    unusedTreasures.add(new Treasure("Necro-comicón", 1, TreasureKind.ONEHAND));
    unusedTreasures.add(new Treasure("Necronomicón",5, TreasureKind.BOTHHANDS));
    unusedTreasures.add(new Treasure("Linterna a 2 manos", 3, TreasureKind.BOTHHANDS));
    unusedTreasures.add(new Treasure("Necro-gnomicón", 2, TreasureKind.ONEHAND));
    unusedTreasures.add(new Treasure("Necrotelecom", 2, TreasureKind.HELMET));
    unusedTreasures.add(new Treasure("Mazo de los antiguos", 3, TreasureKind.ONEHAND));
    unusedTreasures.add(new Treasure("Necro-playboycon", 3, TreasureKind.ONEHAND));
    unusedTreasures.add(new Treasure("Porra preternatural", 2, TreasureKind.ONEHAND));
    unusedTreasures.add(new Treasure("Shogulador", 1, TreasureKind.BOTHHANDS));
    unusedTreasures.add(new Treasure("Varita de atizamiento", 3, TreasureKind.ONEHAND));
    unusedTreasures.add(new Treasure("Tentácula de pega", 2, TreasureKind.HELMET));
    unusedTreasures.add(new Treasure("Zapato deja-amigos", 1, TreasureKind.SHOES));

  }
  private void initMonsterCardDeck(){

       BadConsequence badConsequence1 = new SpecificBadConsequence("Pierdes tu armadura\n visible y otra oculta.", 0 , new ArrayList(Arrays.asList(TreasureKind.ARMOR)),
                new ArrayList(Arrays.asList(TreasureKind.ARMOR)));
         Prize prize1 = new Prize(2,1);
         unusedMonster.add(new Monster("Byakhees de bonanza", 8, badConsequence1, prize1));

         BadConsequence badConsequence2 = new SpecificBadConsequence("Embobados con el lindo primigenio\n te descartas de tu casco visible.", 0 , new ArrayList(Arrays.asList(TreasureKind.HELMET)),
                 new ArrayList(Arrays.asList()) );
         Prize prize2 = new Prize(1,1);
         unusedMonster.add(new Monster("Tenochtitlan", 2, badConsequence2, prize2));

         BadConsequence badConsequence3 = new SpecificBadConsequence("EL primordial bostezo contagioso.\n Pierdes el calzado visible.", 0 , new ArrayList(Arrays.asList(TreasureKind.SHOES)),
                 new ArrayList(Arrays.asList()) );
         Prize prize3 = new Prize(1,1);
         unusedMonster.add(new Monster("El sopor de Dunwich", 2, badConsequence3, prize3));

         BadConsequence badConsequence4 = new SpecificBadConsequence("Te atrapan para llevarte a la fiesta\n y te dejan caer en mitad del vuelo.\n Descarta 1 mano visible  y 1 mano oculta.",
                 0 , new ArrayList(Arrays.asList(TreasureKind.ONEHAND)), new ArrayList(Arrays.asList(TreasureKind.ONEHAND)) );
         Prize prize4 = new Prize(4,1);
         unusedMonster.add(new Monster("Angeles de la noche ibicenca", 14 , badConsequence4, prize4));

         BadConsequence badConsequence5 = new NumericBadConsequence("Pierdes todos tus tesoros visibles.", 0 , 5, 0 );
         Prize prize5 = new Prize(3,1);
         unusedMonster.add(new Monster("El gorron en el umbral", 10 , badConsequence5, prize5));

         BadConsequence badConsequence6 = new SpecificBadConsequence("Pierdes la armadura visible.", 0 ,new ArrayList(Arrays.asList(TreasureKind.ARMOR)),
               new ArrayList(Arrays.asList()));
         Prize prize6 = new Prize(3,1);
         unusedMonster.add(new Monster("H.P. Munchcraft", 6 , badConsequence6, prize6));

         BadConsequence badConsequence7 = new SpecificBadConsequence("Sientes bichos bajo la ropa.\n Descarta tu armadura visible.", 0 ,new ArrayList(Arrays.asList(TreasureKind.ARMOR)),
                 new ArrayList(Arrays.asList()));
         Prize prize7 = new Prize(1,1);
         unusedMonster.add(new Monster("Necrofago", 2 , badConsequence7, prize7));

         BadConsequence badConsequence8 = new NumericBadConsequence("Pierdes 5 niveles\n y 3 tesoros visibles.",5 ,3, 0 );
         Prize prize8 = new Prize(4,2);
         unusedMonster.add(new Monster("El rey de rosa", 13 , badConsequence8, prize8));

         BadConsequence badConsequence9 = new NumericBadConsequence("Toses los pulmones\n y pierdes 2 niveles.",2 ,0, 0 );
         Prize prize9 = new Prize(1,1);
         unusedMonster.add(new Monster("Flecher", 2 , badConsequence9, prize9));

         BadConsequence badConsequence10 = new DeathBadConsequence("Estos monstruos resultan\n bastante superficiales y te aburren mortalmente\n. Estas muerto.",true);
         Prize prize10 = new Prize(2,1);
         unusedMonster.add(new Monster("Los hondos", 8 , badConsequence10, prize10));

         BadConsequence badConsequence11 = new NumericBadConsequence("Pierdes 2 niveles y\n 2 tesoros ocultos." ,2 ,2 ,0 );
         Prize prize11 = new Prize(2,1);
         unusedMonster.add(new Monster("Semillas Cthulhu", 4 , badConsequence11, prize11));

         BadConsequence badConsequence12 = new SpecificBadConsequence("Te intentas escaquear.\n Pierdes una mano visible." ,0 ,new ArrayList(Arrays.asList(TreasureKind.ONEHAND)) ,
                 new ArrayList(Arrays.asList()));
         Prize prize12 = new Prize(2,1);
         unusedMonster.add(new Monster("Dameargo",1 , badConsequence12, prize12));

         BadConsequence badConsequence13 = new NumericBadConsequence("Da mucho asquito.\n Pierdes 3 niveles." ,3 ,0 ,0 );
         Prize prize13 = new Prize(1,1);
         unusedMonster.add(new Monster("Pollipolipo volante",3 , badConsequence13, prize13));

         BadConsequence badConsequence14 = new DeathBadConsequence("No le hace gracia que pronuncien\n mal su nombre. Estas muerto." ,true );
         Prize prize14 = new Prize(3,1);
         unusedMonster.add(new Monster("Yskhtihyssg-Goth",12 , badConsequence14, prize14));

         BadConsequence badConsequence15 = new DeathBadConsequence("La familia te atrapa. Estas muerto." ,true );
         Prize prize15 = new Prize(4,1);
         unusedMonster.add(new Monster("Familia Feliz",1 , badConsequence15, prize15));

         BadConsequence badConsequence16 = new SpecificBadConsequence("La quinta directiva te atrapa\n y te obliga a perder 2 niveles\n y un tesoro de 2 manos visible." ,2 ,
              new ArrayList(Arrays.asList(TreasureKind.BOTHHANDS)), new ArrayList(Arrays.asList()));
         Prize prize16 = new Prize(2,1);
         unusedMonster.add(new Monster("Robotggoth",8 , badConsequence16, prize16));

         BadConsequence badConsequence17 = new SpecificBadConsequence("Te asusta en la noche.\n Pierdes un casco visible" ,0 ,new ArrayList(Arrays.asList(TreasureKind.HELMET)),
                 new ArrayList(Arrays.asList()));
         Prize prize17 = new Prize(1,1);
         unusedMonster.add(new Monster("El espia sordo", 5 , badConsequence17, prize17));

         BadConsequence badConsequence18 = new NumericBadConsequence("Menudo susto te llevas.\n Pierdes 2 niveles y 5 tesoros visibles." ,2 ,5, 0);
         Prize prize18 = new Prize(1,1);
         unusedMonster.add(new Monster("Tongue" ,19 , badConsequence18, prize18));

         BadConsequence badConsequence19 = new SpecificBadConsequence("Te faltan manos para tanta cabeza.\n Pierdes 2 niveles y tus tesoros visibles de las manos." ,2 ,
                 new ArrayList(Arrays.asList(TreasureKind.BOTHHANDS,TreasureKind.ONEHAND,TreasureKind.ONEHAND)),new ArrayList(Arrays.asList()));
         Prize prize19 = new Prize(1,1);
         unusedMonster.add(new Monster("Bicéfalo" ,21 , badConsequence19, prize19));



         BadConsequence badConsequence20 = new SpecificBadConsequence("Pierdes una mano visible" ,0 ,
                 new ArrayList(Arrays.asList(TreasureKind.ONEHAND)),new ArrayList(Arrays.asList()));
         Prize prize20 = new Prize(3,1);
         unusedMonster.add(new Monster("El mal indecible impronunciable." ,10 , badConsequence20, prize20, -2));

         BadConsequence badConsequence21 = new NumericBadConsequence("Pierdes tus tesoros visibles. Jajaja" ,0 ,5,0);
         Prize prize21 = new Prize(2,1);
         unusedMonster.add(new Monster("Testigos Oculares." ,6 , badConsequence21, prize21, +2));

         BadConsequence badConsequence22 = new DeathBadConsequence("Hoy no es tu día de suerte. Mueres." ,true );
         Prize prize22 = new Prize(2,5);
         unusedMonster.add(new Monster("El gran cthulhu.  ",20 , badConsequence22, prize22,+4));

         BadConsequence badConsequence23 = new NumericBadConsequence("Tu gobierno te recorta dos niveles." ,2 ,0,0);
         Prize prize23 = new Prize(2,1);
         unusedMonster.add(new Monster("Serpiente político." ,8 , badConsequence23, prize23, -2));

         BadConsequence badConsequence24 = new SpecificBadConsequence("Pierdes tu casco y armadura visible.\n Pierdes tus manos ocultas." ,0 ,
                 new ArrayList(Arrays.asList(TreasureKind.HELMET,TreasureKind.ARMOR)),new ArrayList(Arrays.asList(TreasureKind.BOTHHANDS,TreasureKind.ONEHAND,TreasureKind.ONEHAND)));
         Prize prize24  = new Prize(1,1);
         unusedMonster.add(new Monster("Felpuggoth." ,2 , badConsequence24, prize24, +5));


         BadConsequence badConsequence25 = new NumericBadConsequence("Pierdes dos niveles." ,2 ,0,0);
         Prize prize25 = new Prize(4,2);
         unusedMonster.add(new Monster("Shoggoth." ,16 , badConsequence25, prize25, -4));

         BadConsequence badConsequence26 = new NumericBadConsequence("Pintalabios negro. Pierdes dos niveles." ,2 ,0,0);
         Prize prize26 = new Prize(1,1);
         unusedMonster.add(new Monster("Lolitaggoth." ,2 , badConsequence26, prize26, +3));


  }

  private void initCultistCardDeck(){

      unusedCultist.add(new Cultist("Sectario", 1));
      unusedCultist.add(new Cultist("Sectario", 2));
      unusedCultist.add(new Cultist("Sectario", 1));
      unusedCultist.add(new Cultist("Sectario", 2));
      unusedCultist.add(new Cultist("Sectario", 1));
      unusedCultist.add(new Cultist("Sectario", 1));
  }

  private void shuffleCultist(){
    Collections.shuffle(unusedCultist);
  }

  private void shuffleTreasures(){
    Collections.shuffle(unusedTreasures);

  }
  private void shuffleMonsters(){
    Collections.shuffle(unusedMonster);
  }

  public Cultist nextCultist(){
    if(unusedCultist.isEmpty()){
      for(Cultist c : usedCultist){
        unusedCultist.add(c);
      }
      shuffleCultist();
    }

    Cultist auxiliar = this.unusedCultist.get(0);
    usedCultist.add(auxiliar);
    unusedCultist.remove(auxiliar);

    return auxiliar;
  }

  public Treasure nextTreasure(){

    if(unusedTreasures.isEmpty()){
      for(Treasure treasures : usedTreasures){
        unusedTreasures.add(treasures);
      }
      shuffleTreasures();
    }

    Treasure auxiliar = this.unusedTreasures.get(0);
    usedTreasures.add(auxiliar);
    unusedTreasures.remove(auxiliar);

    return auxiliar;
  }


  public Monster nextMonster(){

    if(unusedMonster.isEmpty()){

      for(Monster monster : usedMonster){
          unusedMonster.add(monster);
      }
      shuffleMonsters();
    }

    Monster aux = this.unusedMonster.get(0);
    usedMonster.add(aux);
    unusedMonster.remove(aux);

    return aux;
  }

  public void giveCultistBack(Cultist c){
      this.usedCultist.add(c);
  }

  public void giveTreasureBack(Treasure t){
      this.usedTreasures.add(t);
  }

  public void giveMonsterBack(Monster m){
      this.usedMonster.add(m);
  }

  public void initCards(){

      this.initTreasureCardDeck();
      this.initMonsterCardDeck();
      this.initCultistCardDeck();
      shuffleTreasures();
      shuffleMonsters();
      shuffleCultist();
  }

}
